Tools/TestWebKitAPI/Tests/WebKitGLib/TestWebKitWebXR.cpp

```diff
@@ -130,12 +131,116 @@ static void testWebKitWebXRLeaveImmersiveModeAndWaitUntilImmersiveModeChanged(We
+        WebKitXRSessionFeatures consentOptionalFeatures { noFeature };
+        WebKitXRSessionFeatures requiredFeaturesRequested { noFeature };
+        WebKitXRSessionFeatures optionalFeaturesRequested { noFeature };
+    };
+    struct Data {
+        WebViewTest* test { nullptr };
+        Answer answer { Answer::Deny };
+        Result result { };
+
+        void resetResult()
+        {
+            result = Result();
+        }
+    } data { test };
+    typedef gboolean (*PermissionRequestCallback)(WebKitWebView*, WebKitPermissionRequest*, Data*);
+    PermissionRequestCallback permissionRequestCallback = [](WebKitWebView*, WebKitPermissionRequest* request, Data* data) -> gboolean {
+        g_assert_true(WEBKIT_IS_XR_PERMISSION_REQUEST(request));
+        WebKitXRPermissionRequest* xrRequest = WEBKIT_XR_PERMISSION_REQUEST(request);
+
+        data->result.mode = webkit_xr_permission_request_get_session_mode(xrRequest);
+        g_autofree gchar* originStr = webkit_security_origin_to_string(webkit_xr_permission_request_get_security_origin(xrRequest));
+        data->result.origin = String::fromUTF8(originStr);
+        data->result.grantedFeatures = webkit_xr_permission_request_get_granted_features(xrRequest);
+        data->result.consentRequiredFeatures = webkit_xr_permission_request_get_consent_required_features(xrRequest);
+        data->result.consentOptionalFeatures = webkit_xr_permission_request_get_consent_optional_features(xrRequest);
+        data->result.requiredFeaturesRequested = webkit_xr_permission_request_get_required_features_requested(xrRequest);
+        data->result.optionalFeaturesRequested = webkit_xr_permission_request_get_optional_features_requested(xrRequest);
+        data->result.didCallback = true;
+
+        if (data->answer == Answer::Deny)
+            webkit_permission_request_deny(request);
+        else
+            webkit_permission_request_allow(request);
+        return TRUE;
+    };
+
+    g_signal_connect(test->webView(), "permission-request", G_CALLBACK(permissionRequestCallback), &data);
+
+    WebViewTest::NetworkPolicyGuard guard(test, WEBKIT_TLS_ERRORS_POLICY_IGNORE);
+    test->loadHtml("", "https://foo.com/bar");
+    test->waitUntilLoadFinished();
+    test->showInWindow();
+
+    auto testPermissionRequest = [&](StringView mode, StringView options, Answer answer) {
+        auto script = makeString(
+            "async function start() {"
+            "    try {"
+            "        const session = await navigator.xr.requestSession('"_s, mode, "', {"_s, options, "});"
+            "        session.end();"
+            "    } catch (e) {"
+            "    }"
+            "    document.title += 'x';"
+            "}"
+            "start()"_s);
+        data.answer = answer;
+        data.resetResult();
+        test->runJavaScriptAndWaitUntilFinished(script.utf8().data(), nullptr);
+        test->waitUntilTitleChanged();
```

Comment:
Nice, too. I can follow this approach by updating the title if the session fails to start and raises an error -- which it shouldn't unless the environment is not well setup.